### PR TITLE
[Feature] Added feature that allows users to sort the LFD entries by selected fields.

### PR DESCRIPTION
**Pull request by YaroslavOvdii**
@sofiiakvasnevska 

## Issue
https://github.com/Fliplet/fliplet-studio/issues/4999
https://weboo.atlassian.net/browse/OD-20?atlOrigin=eyJpIjoiZDNiZDNhZWQ4MDJlNDA0OWE4NTVjMmJiNDY5YTM4NDciLCJwIjoiaiJ9

## Description
Added feature that allows users to sort the LFD entries by selected fields.

## Screenshots/screencasts
https://share.getcloudapp.com/OAuBLLJj

## Backward compatibility
This change is fully backward compatible.

## Reviewers
@upplabs-alex-levchenko

### CODE REVIEW

File: interface.html [hunk 1]
```diff
@@ -142,6 +142,25 @@
           </div>
         </div>
 
+        <div class="form-group clearfix list-sort disabled hidden">
+          <div class="col-sm-4 control-label">
+            <label>List sort</label>
```
Comment (tonytlwu) [inline interface.html:147]:
```suggestion
            <label>List sorting</label>
```

File: templates/build/news-feed-base.build.hbs [hunk 2]
```diff
@@ -57,6 +57,16 @@
       \{{#if filtersEnabled}}
       <i class="fa fa-sliders" ></i>
       \{{/if}}
+      \{{#if sortEnabled}}
+      <div class="sort-group btn-group">
+        <i class="fa fa-sort-amount-desc" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" aria-hidden="true"></i>
+        <ul class="dropdown-menu list-sort sort-list-news">
```
Comment (tonytlwu) [inline templates/build/news-feed-base.build.hbs:63]: Is it necessary that we use a class name that's specific to the LFD layout? This UI DOM is contained in other containers that have layout specific classes, so it should be sufficient for any kind of selector targeting.

```suggestion
        <ul class="dropdown-menu list-sort">
```

File: js/layout-javascript/news-feed-code.js [hunk 3]
```diff
@@ -157,6 +157,37 @@ DynamicList.prototype.attachObservers = function() {
...
+        'none': 'asc',
+        'asc': 'desc',
+        'desc': 'asc'
+      };
+
+      sortOrder = newSortOrder[oldSortOrder];
+      _this.Utils.DOM.resetSortIcons({ $sortList: $sortList });
+
+      $sortOrderIcon.removeClass(_.values(sortClasses).join(' ')).addClass(sortClasses[sortOrder]);
+      $sortListItem.data('sortOrder', sortOrder);
```
Comment (tonytlwu) [inline js/layout-javascript/news-feed-code.js:184]:
```js
      _this.Utils.DOM.resetSortIcons({ $sortList: $sortList });

      $sortOrderIcon.removeClass(_.values(sortClasses).join(' ')).addClass(sortClasses[sortOrder]);
      $sortListItem.data('sortOrder', sortOrder);
```

These 3 lines should be part of `searchData()`. This way, whenever `searchData()` is executed with valid `sortOrder` and `sortField`, the UI is updated accordingly.

**Review by tonytlwu — CHANGES_REQUESTED** [review]

**Comment by tonytlwu** [comment]
@YaroslavOvdii I looked through how you've implemented sort into the `searchData()` function. There's a flaw that I didn't see before. Can you confirm if my observation is correct?

Scenario:

1. List contains 1 million records
1. A search with filter and search term is executed to return 5 results
1. A sort is applied based on a field and sort order
1. Expected algorithm: The 5 result entries are sorted
1. Current algorithm: The function filters through 1 million records to return 5 results, then sorts them

**Comment by YaroslavOvdii** [comment]
@tonytlwu yes, your scenario is correct. I'll look into it and fix this flaw.

**Comment by YaroslavOvdii** [comment]
@tonytlwu After more close investigating I've discovered that we sort list that was limited.
It will work like that: 

1. List contains 1 million records
2. A search with filter and search term is executed to return 5 results
3. A sort is applied based on a field and sort order
4. Expected algorithm: The 5 result entries are sorted
5. Current algorithm: The 5 result entries are sorted

As you may see on the video the same thing will apply to the list limit, as you may see on the [screencast](https://share.getcloudapp.com/v1u2W0N6)

**Comment by YaroslavOvdii** [comment]
Resolved conflict

File: templates/build/news-feed-base.build.hbs (hunk 2 above)
Comment (YaroslavOvdii) [reply templates/build/news-feed-base.build.hbs:63]: @tonytlwu We can remove this class, all we need is change `resetSortIcons` function.
And if we remove this class that is specific for layout, may I suggest moving this [function](https://github.com/Fliplet/fliplet-widget-dynamic-lists/pull/426/files/2a8ab1169c25d50f7b90edf4b467dd4010ca9c98#diff-f578a21548d18676ae0271249fa231eeR161) to the Utils.js and made it common to all layouts?

Comment (tonytlwu) [reply templates/build/news-feed-base.build.hbs:63]: @YaroslavOvdii I don't mind. Go ahead and make it DRY.

**Comment by tonytlwu** [comment]
@YaroslavOvdii [Your screencast](https://share.getcloudapp.com/v1u2W0N6) showed you using the **Limit entries** feature. The scenarios you and I outlined does not mention the feature being used. Specifically:

1. List contains 10 records
1. Render the full list
1. Apply a filter using LFD UI to return a subset, e.g. 2 entries
1. Apply a sort based on a field and sort order
1. Expected algorithm: The sort is applied directly to the filtered result, e.g. 2 entries
1. Current algorithm: The function filters through all 10 records to return the subset, e.g. 2 entries, then sorts them

**Comment by YaroslavOvdii** [comment]
@tonytlwu Yes. you are right at my screencast I showed the example with limited entries but it worked the same for filtered entries as you can see [here](https://share.getcloudapp.com/WnuJ8ZZG).
The reason why it is working like that because we take data after it was filtered, limited and add summary data in this function `addSummaryData` and I'm sorting data that we will show to the user not all of them.

**Comment by tonytlwu** [comment]
@YaroslavOvdii Your video has shown that, when user sorts a filtered list, between:

```js
_this.modifiedListItems = _this.addSummaryData(searchedData);
```

...and

```js
_this.modifiedListItems = _this.Utils.Records.sortByField(data);
```

...`_this.modifiedListItems` has gone from 12 entries to 4 entries. I assume this is because `searchedData` contains 4 entries.

`searchedData` comes from the result of `_this.Utils.Records.runSearch()`. Can you confirm, given the list is already filtered, how many entries/records are passed to the `_this.Utils.Records.runSearch()` function, and therefore how many loops does `_this.Utils.Records.runSearch()` go through to produce the `searchedData` array when you are sorting the filtered list?

**Comment by YaroslavOvdii** [comment]
@tonytlwu From what I've researched current behavior is:
1. We have a total of 3 items on the list.
2. We start a search and left the only item.
3. We start a sort.
3.1 It began `the _this.Utils.Records.runSearch()` function which is start a search once again and will search in all 3 items.

Do I get it right that we don't want 3.1 to be happening?

**Comment by tonytlwu** [comment]
@YaroslavOvdii If you have 10 million records, filtered down to 2 entries. If you run a sort, you want to sort to happen directly based on the 2 entries, not loop through 10 million records again to produce the 2 entries again before sorting it.

So no, we don't want sorting to happen as you've described in 3.1.

Also, the sorting can happen without re-rendering and can just shuffle the onscreen DOM elements. This way, the social features (bookmark/like/comment) don't need to reinitialise again.

**Comment by YaroslavOvdii** [comment]
@tonytlwu Yes, I get it now. I'll try to implement this and return to you with a solution.

**Comment by YaroslavOvdii** [comment]
@tonytlwu Could you please help me with one issue?
I was able to do as you said and made sorting without a re-rendering, but I've encounter another issue.
After I sort the list, the bookmark became inactive, and only after I check the bookmark in the detail view that came back to life.
[Screencast](https://share.getcloudapp.com/lluYA1g1)
Altered sort function that I use for this.
```
  function sortByField(options) {
    // If user doesn't set sorting do nothing
    if (!options.sortField) {
      return options.records;
    }

    var $container = options.$container;
    var $list = $container.find(options.listContainer);
    var listItems = $list.children();
    var sortedItemsList = [];
    var records = _.clone(options.records);
    var isSortAsc = options.sortOrder === 'asc';
    var sortField = options.sortField;
    var startsWithAlphabet = /^[A-Z,a-z]/;
    var sortType = getFieldType(records[0][sortField]);
  
    records.sort(function(a, b) {
      var aValue = a.data[sortField];
      var bValue = b.data[sortField];
      
      switch (sortType) {
        case 'string': 
          /**
           * By adding '{' to the start of the string we are pushing that string to the end the sort order,
           * and when we are adding '}' we push it even further to the end
          */
          aValue = aValue ? aValue.toUpperCase() : '}';
          bValue = bValue ? bValue.toUpperCase() : '}';

          if (!startsWithAlphabet.test(bValue)) {
            bValue = '{' + bValue;
          }

          if (!startsWithAlphabet.test(aValue)) {
            aValue = '{' + aValue;
          }
          
          if (aValue < bValue) {
            return isSortAsc ? -1 : 1;
          }
          if (aValue > bValue) {
            return isSortAsc ? 1 : -1;
          }

          return 0;
        case 'number':
          if (!parseInt(aValue, 10)) {
            return isSortAsc ? 1 : -1;
          }

          if (!parseInt(bValue, 10)) {
            return isSortAsc ? -1 : 1;
          }

          if (isSortAsc) {
            return aValue - bValue;
          }

          return bValue - aValue;
        case 'time':
          if (!aValue) {
            return isSortAsc ? 1 : -1;
          }

          if (!bValue) {
            return isSortAsc ? -1 : 1;
          }

          if (aValue < bValue) {
            return isSortAsc ? -1 : 1;
          }

          if (aValue > bValue) {
            return isSortAsc ? 1 : -1;
          }

          return 0;
        case 'date':
          aValue = moment(aValue).valueOf();
          bValue = moment(bValue).valueOf();

          if (!aValue) {
            return isSortAsc ? 1 : -1;
          }

          if (!bValue) {
            return isSortAsc ? -1 : 1;
          }

          if (isSortAsc) {
            return aValue - bValue;
          }

          return bValue - aValue;
         default: 
          if (isSortAsc) {
            return aValue - bValue;
          }

          return bValue - aValue;
      }
    });

    listItems.each(function(index) {
      var $listItem = $(listItems[index]);
      var itemId = $listItem.data('entry-id');
      var itemSortedIndex = records.findIndex(function(record) {
        return record.id === itemId;
      });

      if (itemSortedIndex !== -1) {
        sortedItemsList[itemSortedIndex] = $listItem;
      }
      
    });

    $list.html(sortedItemsList);
  }
```

And the function that is calling it:
```
    .on('click', '.sort-group .list-sort li', function (e) {
      e.stopPropagation();

      var $sortListItem = $(e.currentTarget);
      var oldSortOrder = $sortListItem.data('sortOrder');
      var sortOrder;
      var sortField = $sortListItem.data('sortField');;
      var $sortOrderIcon = $sortListItem.find('i');
      var $sortList = _this.$container.find('.sort-list-directory li');
      var sortClasses = {
        'none': 'fa-sort',
        'asc': 'fa-sort-asc',
        'desc': 'fa-sort-desc'
      };
      var newSortOrder = {
        'none': 'asc',
        'asc': 'desc',
        'desc': 'asc'
      };

      sortOrder = newSortOrder[oldSortOrder];
      _this.Utils.DOM.resetSortIcons({ $sortList: $sortList });

      $sortOrderIcon.removeClass(_.values(sortClasses).join(' ')).addClass(sortClasses[sortOrder]);
      $sortListItem.data('sortOrder', sortOrder);

      _this.Utils.Records.sortByField({
        $container: _this.$container,
        listContainer: '#small-card-list-wrapper-' + _this.data.id,
        records: _this.searchedListItems,
        sortOrder: sortOrder,
        sortField: sortField
      });
    })
```

**Comment by tonytlwu** [comment]
@YaroslavOvdii See if this makes sense to you https://jsfiddle.net/tonytlwu/d8e9gvxc/ Essentially your code is following the pattern of `#one` but that removes the event listener and you should follow the pattern in `#two`.

**Comment by YaroslavOvdii** [comment]
@tonytlwu Thanks your example helped a lot.
Now we sort only items that were filtered, searched, or bookmarked.
And we do not re-render the layout but simple sort HTML elements.

File: js/utils.js [hunk 4]
```diff
@@ -1132,7 +1132,40 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+    });
+  }
+
+  function sortItems(options) {
+    var records = sortByField(options);
+
+    sortHTMLElements({
+      $layoutContainer: options.$container,
+      listContainer: options.listContainer,
+      records: records
```
Comment (tonytlwu) [inline js/utils.js:1144]: For clarity

```suggestion
      sortedRecords: sortedRecords
```

File: js/utils.js [hunk 5]
```diff
@@ -1529,7 +1562,7 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
       updateFiles: updateRecordFiles,
       prepareData: prepareRecordsData,
       addComputedFields: addRecordsComputedFields,
-      sortByField: sortByField
+      sortByField: sortItems
```
Comment (tonytlwu) [inline js/utils.js:1565]:
```suggestion
      sortByField: sortRecordsByField
```

File: js/utils.js [hunk 6]
```diff
@@ -1132,7 +1132,40 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+    });
+  }
+
+  function sortHTMLElements(options) {
+    var sortedItemsList = [];
+    var $list = options.$layoutContainer.find(options.listContainer);
+    var listItems = $list.children();
+
+    listItems.each(function(index) {
+      var $listItem = $(listItems[index]);
```
Comment (tonytlwu) [inline js/utils.js:1154]:
```suggestion
      var $listItem = $(this);
```

File: js/utils.js [hunk 7]
```diff
@@ -1132,7 +1132,40 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+      records: records
+    });
+  }
+
+  function sortHTMLElements(options) {
+    var sortedItemsList = [];
+    var $list = options.$layoutContainer.find(options.listContainer);
+    var listItems = $list.children();
+
+    listItems.each(function(index) {
```
Comment (tonytlwu) [inline js/utils.js:1153]:
```suggestion
    listItems.each(function() {
```

File: js/utils.js [hunk 8]
```diff
@@ -1132,7 +1132,40 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+    sortHTMLElements({
+      $layoutContainer: options.$container,
+      listContainer: options.listContainer,
+      records: records
+    });
+  }
+
+  function sortHTMLElements(options) {
+    var sortedItemsList = [];
+    var $list = options.$layoutContainer.find(options.listContainer);
```
Comment (tonytlwu) [inline js/utils.js:1150]: The way you've used `.detach()` in a loop at the moment will result in each list item being detached one after another. This is CPU heavy for the rendering engine.

Instead, detach the parent container of the list items once, update its content (like you're currently doing), and then put the `$list` element back.

```suggestion
    var $list = options.$layoutContainer.find(options.listContainer).detach();
```

**Review by tonytlwu — CHANGES_REQUESTED** [review]

**Comment by YaroslavOvdii** [comment]
@tonytlwu Added requested changes

File: js/utils.js [hunk 9]
```diff
@@ -1025,6 +1025,170 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+  }
+
+  function sortHTMLElements(options) {
+    var sortedItemsList = [];
+    var $list = options.$layoutContainer.find(options.listContainer).detach();
+    var $listItems = $list.children();
+
+    $listItems.each(function() {
+      var $listItem = $(this);
+      var itemId = $listItem.data('entry-id');
```
Comment (tonytlwu) [inline js/utils.js:1155]: Skip the jQuery object wrapping so you can do this. We're looking at a performance-sensitive operation here so it's best to cut down on unnecessary operations.

```suggestion
      var itemId = parseInt(this.dataset.entryId, 10);
```

File: js/utils.js [hunk 10]
```diff
@@ -1025,6 +1025,170 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+      var itemSortedIndex = options.records.findIndex(function(record) {
```
Comment (tonytlwu) [inline js/utils.js:1156]: Are you sure this works?

```suggestion
      var itemSortedIndex = options.sortedRecords.findIndex(function(record) {
```

File: js/utils.js [hunk 11]
```diff
@@ -1025,6 +1025,170 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+        return record.id === itemId;
+      });
+
+      if (itemSortedIndex !== -1) {
+        sortedItemsList[itemSortedIndex] = $listItem;
+      }
+      
+    });
+
+    $list.html(sortedItemsList);
```
Comment (tonytlwu) [inline js/utils.js:1166]: Are you sure this works? It looks like you detached `$list` but never put it back into the document.

**Review by tonytlwu — CHANGES_REQUESTED** [review]
Here's a scenario we might not have thought of before.

GIVEN sorting is active using a specific field
  WHEN user executes a filter or search
    THEN the filter or search result should keep the sorting order

Can you confirm if the scenario above is working? We need this working.

File: js/layout-javascript/news-feed-code.js [hunk 12]
```diff
@@ -1639,7 +1675,10 @@ DynamicList.prototype.searchData = function(options) {
       _this.$container.find('.hidden-search-controls').addClass('active');
       _this.$container.find('.hidden-search-controls')[searchedData.length || truncated ? 'removeClass' : 'addClass']('no-results');
 
-      if (!_this.data.forceRenderList && searchedData.length && !_.xorBy(searchedData, _this.searchedListItems, 'id').length) {
+      if (!_this.data.forceRenderList
```
[the remark this answers is not in the document]
Comment (YaroslavOvdii) [reply js/layout-javascript/news-feed-code.js:1678]: Old code, I'll remove it.

File: js/layout-javascript/news-feed-code.js [hunk 13]
```diff
@@ -1566,6 +1600,8 @@ DynamicList.prototype.searchData = function(options) {
 
   var _this = this;
   var value = _.isUndefined(options.value) ? _this.searchValue : ('' + options.value).trim();
+  var sortField = options.sortField || '';
+  var sortOrder = options.sortOrder || 'asc';
```
[the remark this answers is not in the document]
Comment (YaroslavOvdii) [reply js/layout-javascript/news-feed-code.js:1604]: We do not need it anymore, I'll remove it.

File: js/utils.js (hunk 11 above)
Comment (YaroslavOvdii) [reply js/utils.js:1173]: Fixed

File: js/utils.js (hunk 10 above)
Comment (YaroslavOvdii) [reply js/utils.js]: Fixed

File: js/layout-javascript/news-feed-code.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (YaroslavOvdii) [reply js/layout-javascript/news-feed-code.js]: Fixed

File: js/layout-javascript/news-feed-code.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (YaroslavOvdii) [reply js/layout-javascript/news-feed-code.js:1679]: Fixed

**Comment by YaroslavOvdii** [comment]
> Here's a scenario we might not have thought of before.
GIVEN sorting is active using a specific field
WHEN user executes a filter or search
THEN the filter or search result should keep the sorting order
Can you confirm if the scenario above is working? We need this working.

This scenario will work now.

File: templates/build/news-feed-base.build.hbs (hunk 2 above)
Comment (tonytlwu) [reply templates/build/news-feed-base.build.hbs:63]: This still needs action

File: js/utils.js [hunk 14]
```diff
@@ -1025,6 +1025,178 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+      var $listItem = $(this);
+      var itemId = parseInt($listItem.data('entry-id'), 10);
+      var itemSortedIndex = options.sortedRecords.findIndex(function(record) {
+        return record.id === itemId;
+      });
+
+      if (itemSortedIndex !== -1) {
+        sortedItemsList[itemSortedIndex] = $listItem;
+      }
+
```
Comment (tonytlwu) [inline js/utils.js:1170]:
```suggestion
```

File: js/utils.js [hunk 15]
```diff
@@ -1025,6 +1025,178 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+    var sortedItemsList = [];
+    var $list = $(options.listContainer);
+    var $prevElement = $list.prev();
+    var $detachedList = $list.detach();
+    var $listItems = $detachedList.find('.small-card-list-item').detach();
+
+    $listItems.each(function() {
+      var $listItem = $(this);
+      var itemId = parseInt($listItem.data('entry-id'), 10);
+      var itemSortedIndex = options.sortedRecords.findIndex(function(record) {
```
Comment (tonytlwu) [inline js/utils.js:1163]: Not supported in IE.

```suggestion
      var itemSortedIndex = _.findIndex(options.sortedRecords, function(record) {
```

File: js/utils.js [hunk 16]
```diff
@@ -1025,6 +1025,178 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+      sortedRecords: sortedRecords
+    });
+  }
+
+  function sortHTMLElements(options) {
+    var sortedItemsList = [];
+    var $list = $(options.listContainer);
+    var $prevElement = $list.prev();
+    var $detachedList = $list.detach();
+    var $listItems = $detachedList.find('.small-card-list-item').detach();
```
Comment (tonytlwu) [inline js/utils.js:1158]: The `.small-card-list-item` class suggests this probably only works with the directory layout.

File: js/layout-javascript/news-feed-code.js [hunk 17]
```diff
@@ -1672,6 +1709,13 @@ DynamicList.prototype.searchData = function(options) {
       $('#news-feed-list-wrapper-' + _this.data.id).html('');
 
       _this.modifiedListItems = _this.addSummaryData(searchedData);
+      _this.modifiedListItems = _this.Utils.Records.sortByField({
+        records: _this.modifiedListItems,
+        sortOrder: _this.sortOrder,
+        sortField: _this.sortField,
+        onlyRecords: true
```
Comment (tonytlwu) [inline js/layout-javascript/news-feed-code.js:1716]:
```suggestion
        sortHTMLElements: false
```

File: js/utils.js [hunk 18]
```diff
@@ -1025,6 +1025,178 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+
+          return bValue - aValue;
+      }
+    });
+  }
+
+  function sortRecordsByField(options) {
+    var sortedRecords = sortByField(options);
+
+    if (options.onlyRecords) {
```
Comment (tonytlwu) [inline js/utils.js:1141]:
```suggestion
    if (options.sortHTMLElements === false) {
```

File: js/utils.js [hunk 19]
```diff
@@ -1025,6 +1025,178 @@ Fliplet.Registry.set('dynamicListUtils', (function () {
...
+      return sortedRecords;
+    }
+
+    sortHTMLElements({
+      $layoutContainer: options.$container,
+      listContainer: options.listContainer,
+      listHolder: options.listHolder,
+      sortedRecords: sortedRecords
+    });
```
Comment (tonytlwu) [inline js/utils.js:1150]: Ensure consistent function output.

```suggestion
    });

    return sortedRecords;
```

**Review by tonytlwu — CHANGES_REQUESTED** [review]

**Comment by YaroslavOvdii** [comment]
@tonytlwu added suggestions

File: js/utils.js (hunk 15 above)
Comment (tonytlwu) [reply js/utils.js:1165]: This still needs action

Comment (YaroslavOvdii) [reply js/utils.js]: Fixed

File: js/layout-javascript/news-feed-code.js [hunk 20]
```diff
@@ -165,6 +167,40 @@ DynamicList.prototype.attachObservers = function() {
         console.error(error);
       });
     })
+    .on('click', '.sort-group .list-sort li', function (e) {
```
Comment (tonytlwu) [inline js/layout-javascript/news-feed-code.js:170]:
```suggestion
    .on('click', '.sort-group .list-sort li', function(e) {
```

File: js/layout-javascript/news-feed-code.js [hunk 21]
```diff
@@ -165,6 +167,40 @@ DynamicList.prototype.attachObservers = function() {
...
+      e.stopPropagation();
+
+      var $sortListItem = $(e.currentTarget);
+      var oldSortOrder = $sortListItem.data('sortOrder');
+      var $sortOrderIcon = $sortListItem.find('i');
+      var $sortList = _this.$container.find('.list-sort li');
+      var sortClasses = {
+        'none': 'fa-sort',
```
Comment (tonytlwu) [inline js/layout-javascript/news-feed-code.js:178]: Single quote not needed.

```suggestion
        none: 'fa-sort',
```

File: js/layout-javascript/news-feed-code.js [hunk 22]
```diff
@@ -165,6 +167,40 @@ DynamicList.prototype.attachObservers = function() {
...
+        'asc': 'fa-sort-asc',
+        'desc': 'fa-sort-desc'
+      };
+      var newSortOrder = {
+        'none': 'asc',
+        'asc': 'desc',
+        'desc': 'asc'
+      };
+
+      _this.sortOrder = newSortOrder[oldSortOrder];
```
Comment (tonytlwu) [inline js/layout-javascript/news-feed-code.js:188]: Don't need the `newSortOrder` or `oldSortOrder` variables.

```suggestion
      _this.sortOrder = $sortListItem.data('sortOrder') === 'asc' ? 'desc' : 'asc';
```

File: js/layout-javascript/news-feed-code.js [hunk 23]
```diff
@@ -165,6 +167,40 @@ DynamicList.prototype.attachObservers = function() {
...
+      _this.sortField = $sortListItem.data('sortField');;
+      _this.Utils.DOM.resetSortIcons({ $sortList: $sortList });
+
+      $sortOrderIcon.removeClass(_.values(sortClasses).join(' ')).addClass(sortClasses[_this.sortOrder]);
+      $sortListItem.data('sortOrder', _this.sortOrder);
+
+      _this.Utils.Records.sortByField({
+        $container: _this.$container,
+        listContainer: '#news-feed-wrapper-' + _this.data.id,
```
Comment (tonytlwu) [inline js/layout-javascript/news-feed-code.js:197]: We already have a `options.$container` being passed. Why do we need to pass this selector string just for it to be used to create a new `$(options.listContainer)` object?

**Review by tonytlwu — CHANGES_REQUESTED** [review]

File: js/layout-javascript/news-feed-code.js (hunk 23 above)
Comment (YaroslavOvdii) [reply js/layout-javascript/news-feed-code.js:197]: @tonytlwu the `listContainer` we are using to cut the items wrapper, so we could sort items in it and past it to the same place where it was.

Comment (tonytlwu) [reply js/layout-javascript/news-feed-code.js:191]: I see. Let's switch it up a little bit.

```suggestion
        $listContainer: $('#news-feed-wrapper-' + _this.data.id),
```

**Review by tonytlwu — CHANGES_REQUESTED** [review]

**Review by tonytlwu — APPROVED** [review]
@sofiiakvasnevska This is ready for testing

Risk: Critical
Affected: All LFD layouts except _Featured List_
Verify: All LFD render and list features work as expected

Squash and merge: Adds sorting to LFD

**Pull request merged**